### PR TITLE
Templating: Optionally save interpolated expressions when replacing variables in a string

### PR DESCRIPTION
**Pull request by ifrost**
*Background*

#62926 introduced a new method `getAllVariablesInTarget` that was not working with dashboard variables. The change was reverted in #65315. We've discussed possible options to support getting interpolation results of `replace()` method required by #61782. This PR shows a suggested solution.

*Implementation*

`replace()` method takes an optional argument that will save the results of interpolation to a map.

### CODE REVIEW

File: public/app/features/templating/template_srv.ts [hunk 1]
```diff
@@ -291,63 +301,96 @@ export class TemplateSrv implements BaseTemplateSrv {
...
+      });
+    }
 
-        if (value !== null && value !== undefined) {
-          if (scopedVars[variableName].skipFormat) {
-            fmt = undefined;
-          }
+    return this._replaceInVariableRegex(target, format, (match, variableName, fieldPath, fmt) => {
+      return this.evaluateVariableExpression(match, variableName, fieldPath, fmt, scopedVars);
+    });
```
Comment (ifrost) [inline public/app/features/templating/template_srv.ts:316]: @gelicia, in your branch (https://github.com/grafana/grafana/compare/kristina/correlations-log-line?expand=1) it's slightly different with evaluation happening twice. Do we need it for #61782?

File: packages/grafana-runtime/src/services/templateSrv.ts [hunk 2]
```diff
@@ -16,7 +34,12 @@ export interface TemplateSrv {
   /**
    * Replace the values within the target string.  See also {@link InterpolateFunction}
    */
-  replace(target?: string, scopedVars?: ScopedVars, format?: string | Function): string;
+  replace(
+    target?: string,
+    scopedVars?: ScopedVars,
+    format?: string | Function,
+    interpolations?: VariableInterpolation[]
```
Comment (dprokop) [inline packages/grafana-runtime/src/services/templateSrv.ts:41]: Interesting idea with recording the interpolations. I don't think we often use functions that mutate arguments, so I suggest providing a bold doc comment about why and how to use it.

Comment (torkelo) [reply packages/grafana-runtime/src/services/templateSrv.ts:41]: @dprokop yea, agree a good doc comment here would be good. 

or we add another function that returns `{ result: string, interpolations: VariableInterpolation[] }`, but not sure what it should be named, and adds more duplication/choice

File: public/app/features/templating/template_srv.ts (hunk 1 above)
Comment (gelicia) [reply public/app/features/templating/template_srv.ts]: You can just merge this, it's most likely that I missed something and I can figure out what we need.

File: packages/grafana-runtime/src/services/templateSrv.ts (hunk 2 above)
Comment (ifrost) [reply packages/grafana-runtime/src/services/templateSrv.ts:47]: Added a comment in 73cf2e9b26cceb27cd893e6c3b5c71be3c2d790c

**Review by gelicia — APPROVED** [review]

File: packages/grafana-runtime/src/services/templateSrv.ts (hunk 2 above)
Comment (ifrost) [reply packages/grafana-runtime/src/services/templateSrv.ts:47]: @torkelo / @dprokop - could you have another look? Does it look good?

**Review by dprokop — APPROVED** [review]

**Review by torkelo — APPROVED** [review]

**Pull request merged**